verify_rescue_musig added to EddsaSignature

StanislavBreadless opened the pull request:
Added rescue musig verification that adds constraints to the constraint system and a test that shows that it works properly.

Also I added some utilities without which the verification algorithm wouldn't work,

StanislavBreadless committed:
verify_rescue_musig added to EddsaSignature

burdges commented:
I'm curious: What doers this do?

shamatar commented:
This is a backport from zksync for MuSig with Rescue as a hash function, but it doesn't look proper yet. In any case, this is work of our summer interns and will be reviewed.

ly0va committed:
Add functions for rescue musig that operate on field elements

ly0va committed:
Merge remote-tracking branch 'stas/verify_rescue_musig' into plonk_gadgets

StanislavBreadless committed:
Add verify_field_rescue_musig for circuits and tests for it

ly0va committed:
Fix some warnings from cargo clippy

- Remove unnecessary returns
- Remove .clone() on Copy types
- Remove boolean comparisons (if flag == true -> if flag)
- Use .is_empty() instead of comparing .len() to 0
- Use augmented assignments when possible
- Struct { x: x } -> Struct { x }
- .map(|x| f(x)) -> .map(f)
- .map(|x| x.clone()) -> .cloned()
- .map(|x| *x) -> .copied()
- .iter().cloned().collect::<Vec<_>>() -> .to_vec()
- Other minor improvements

ly0va committed:
Merge branch 'verify_rescue_musig' of github.com:StanislavBreadless/franklin-crypto into plonk_gadgets
